### PR TITLE
sql/index: auto delete corrupted indexes

### LINKED ISSUE #220: Auto-detect when pilosa index is corrupt and remove it

### PR DESCRIPTION
**Pull request by erizocosmico**
Closes #220

### CODE REVIEW

File: sql/index/pilosa/driver.go [hunk 1]
```diff
@@ -116,6 +105,50 @@ func (d *Driver) LoadAll(db, table string) ([]sql.Index, error) {
 	return indexes, err
 }
 
+var errCorruptIndex = errors.NewKind("the index in %q is corrupt")
+
+func (d *Driver) loadIndex(path string) (sql.Index, error) {
+	cfg, err := index.ReadConfigFile(path)
```
Comment (ajnavarro) [inline sql/index/pilosa/driver.go]: Maybe the error is permission denied.

**Comment by erizocosmico** [comment]
Sounds good to me

**Comment by kuba--** [comment]
we can reuse an empty bolt db file. moreover bolt set a lock till is opened

**Comment by erizocosmico** [comment]
@ajnavarro an issue I see with this approach:

 We can know if the index was created or not by checking the `CREATED` file.
But if it's not created, the index is corrupt and we have to delete it. Then we need to read the config, which may fail. What does it happen if reading the config fails? If we don't read the config we can't delete the index because we need that data for removing the pilosa stuff.

After all, it's the same logic as before, adding the extra step of having to check for the `CREATED` file, which is overcomplicating the thing. The only part that is simplified is the part of writing the file at the end of the index save.

I mean, we can use `CREATE`, but we still need to think what to do with config read files and it's basically the same logic with extra files.

Create should not be idempotent because an index cannot be overwritten, only deleted and recreated, right? Or am I missing something?

**Comment by ajnavarro** [comment]
@erizocosmico why do you need to read the config file if the index is corrupt? Just ignore it on LoadAll method, and when someone call again Create, override all the files.

**Comment by erizocosmico** [comment]
So you mean we should leave everything there and wait for recreation? Wouldn't it be better to delete all the garbage that a partial index may have written (which potentially can be a lot)?

**Comment by kuba--** [comment]
So that's why I wanted to reuse boltDB file as a lock. It can give us double-check.
- if file exists
- can you open (if not most likely lock is still set)

**Comment by erizocosmico** [comment]
Changed the logic

File: sql/index/pilosa/driver.go [hunk 2]
```diff
@@ -10,6 +10,7 @@ import (
 	"strings"
 
 	pilosa "github.com/pilosa/go-pilosa"
+	"github.com/sirupsen/logrus"
```
Comment (kuba--) [inline sql/index/pilosa/driver.go]: go-log?

File: sql/index/pilosa/driver.go [hunk 3]
```diff
@@ -116,6 +105,37 @@ func (d *Driver) LoadAll(db, table string) ([]sql.Index, error) {
...
+	}
+
+	if ok {
+		log := logrus.WithFields(logrus.Fields{
+			"err":  err,
+			"path": path,
+		})
+		log.Warn("could not read index file, index is corrupt and will be deleted")
+
+		if err := os.RemoveAll(path); err != nil {
```
Comment (kuba--) [inline sql/index/pilosa/driver.go]: So when the file exists we just delete it, not checking how long it exists?
it means that we can save and load, over and over

File: sql/index/pilosa/driver.go (hunk 2 above)
Comment (erizocosmico) [reply sql/index/pilosa/driver.go]: not yet inside go-mysql-server. Last news i had was to wait until a logging guide came out

File: sql/index/pilosa/driver.go (hunk 3 above)
Comment (erizocosmico) [reply sql/index/pilosa/driver.go]: This is only used when the app starts

**Comment by ajnavarro** [comment]
@kuba-- PTAL

**Pull request merged**